### ISSUE TITLE
Imprecise deserialization of F1000 in Release build, caused by compiler's optimization

### ISSUE THREAD

**Issue by asrelo** [minetest/minetest#14879]
### Minetest version

```true
Minetest 5.9.0-dev-a7a719261 (Windows)
Using LuaJIT 2.1.0-beta3
Built by MSVC 19.40
Running on Windows/10.0.19041 x86_64
BUILD_TYPE=Release
RUN_IN_PLACE=0
USE_CURL=1
USE_GETTEXT=0
USE_SOUND=1
STATIC_SHAREDIR="."
```

### Operating system and version

Windows 10.0.19041

### CPU model

Intel Core i5-7200U

### Summary

Before 42cf5e972d1e27a92048712bc79806e1a088b96c, deserialization of F1000 (fixed-point number with 3 decimal digits after mark) was being done as multiplication by `(1.f / 1000.f)`. The referenced commit says that multiplication by `(1./1000.)` introduces error by 1 ULP (because floating point), and fixes it by changing it to division by `1000.`.

On MSVC*, in Release build, the flag `/fp:fast` is applied in `src/CMakeLists.txt` to enable optimizations for floating point operations. This causes the compiler to generate code with fp multiplication by `0.001` instead of division, and reintroduces the error exactly by 1 ULP.

This causes the unit test `TestSerialization::testStreamRead` to fail in Release build by this compiler.

These problems do not happen when I remove `/fp:fast`.

*MSVC version: not entirely sure if I called the correct executable but `Microsoft (R) C/C++ Optimizing Compiler Version 19.40.33812 for x86`

### Steps to reproduce

1. Obtain a build in Release configuration by MSVC.

2. Run `minetest.exe --run-unittests --test-module TestSerialization`

```
======== Testing module TestSerialization
[PASS] testSerializeString - 0ms
[PASS] testDeSerializeString - 1ms
[PASS] testSerializeLongString - 0ms
[PASS] testDeSerializeLongString - 0ms
[PASS] testSerializeJsonString - 1ms
Test assertion failed: readF1000(is) == 53.534f
    at test_serialization.cpp:277
[FAIL] testStreamRead - 2ms
[PASS] testStreamWrite - 0ms
[PASS] testFloatFormat - 902ms
======== Module TestSerialization failed (1 failures / 8 tests) - 920ms
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Unit Test Results: FAILED
    1 / 8 failed tests.
    Testing took 922ms.
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

**Comment by asrelo** [comment]
Release build with `/fp:fast`. `readF1000` is inlined; see `mulss` instruction at `00007FF77E2C2C06`. Value at `07FF77E5A0034h` is `6f12833a` (32-bit little-endian `0.001000000047497451305389404296875`).

<details>
<summary>Disassembled code from the unit test</summary>

```
	UASSERT(readF1000(is) == 53.534f);
00007FF77E2C2BAF  mov         dword ptr [rbp+0D0h],edi  
00007FF77E2C2BB5  mov         r8d,4  
00007FF77E2C2BBB  lea         rdx,[rbp+0D0h]  
00007FF77E2C2BC2  lea         rcx,[rbp-40h]  
00007FF77E2C2BC6  call        qword ptr [__imp_std::basic_istream<char,std::char_traits<char> >::read (07FF77E462658h)]  
00007FF77E2C2BCC  movzx       ecx,byte ptr [rbp+0D0h]  
00007FF77E2C2BD3  shl         ecx,8  
00007FF77E2C2BD6  movzx       eax,byte ptr [rbp+0D1h]  
00007FF77E2C2BDD  or          ecx,eax  
00007FF77E2C2BDF  shl         ecx,8  
00007FF77E2C2BE2  movzx       eax,byte ptr [rbp+0D2h]  
00007FF77E2C2BE9  or          ecx,eax  
00007FF77E2C2BEB  shl         ecx,8  
00007FF77E2C2BEE  movzx       eax,byte ptr [rbp+0D3h]  
00007FF77E2C2BF5  or          ecx,eax  
00007FF77E2C2BF7  movd        xmm0,ecx  
00007FF77E2C2BFB  cvtdq2ps    xmm0,xmm0  
00007FF77E2C2BFE  movss       xmm6,dword ptr [__real@3a83126f (07FF77E5A0034h)]  
00007FF77E2C2C06  mulss       xmm0,xmm6  
00007FF77E2C2C0A  ucomiss     xmm0,dword ptr [__real@425622d1 (07FF77E5A0798h)]  
00007FF77E2C2C11  jne         TestSerialization::testStreamRead+0D35h (07FF77E2C34E5h)  
```

</details>

Release build without `/fp:fast`. `readF1000` is inlined; see `divss` instruction at `00007FF68FF04E06`. Value at `07FF6901E185Ch` is `00007a44` (32-bit little-endian float `+1000.` exactly).

<details>
<summary>Disassembled code from the unit test</summary>

```
	UASSERT(readF1000(is) == 53.534f);
00007FF68FF04DAF  mov         dword ptr [rbp+0D0h],edi  
00007FF68FF04DB5  mov         r8d,4  
00007FF68FF04DBB  lea         rdx,[rbp+0D0h]  
00007FF68FF04DC2  lea         rcx,[rbp-40h]  
00007FF68FF04DC6  call        qword ptr [__imp_std::basic_istream<char,std::char_traits<char> >::read (07FF6900A3658h)]  
00007FF68FF04DCC  movzx       ecx,byte ptr [rbp+0D0h]  
00007FF68FF04DD3  shl         ecx,8  
00007FF68FF04DD6  movzx       eax,byte ptr [rbp+0D1h]  
00007FF68FF04DDD  or          ecx,eax  
00007FF68FF04DDF  shl         ecx,8  
00007FF68FF04DE2  movzx       eax,byte ptr [rbp+0D2h]  
00007FF68FF04DE9  or          ecx,eax  
00007FF68FF04DEB  shl         ecx,8  
00007FF68FF04DEE  movzx       eax,byte ptr [rbp+0D3h]  
00007FF68FF04DF5  or          ecx,eax  
00007FF68FF04DF7  movd        xmm0,ecx  
00007FF68FF04DFB  cvtdq2ps    xmm0,xmm0  
00007FF68FF04DFE  movss       xmm6,dword ptr [__real@447a0000 (07FF6901E185Ch)]  
00007FF68FF04E06  divss       xmm0,xmm6  
00007FF68FF04E0A  ucomiss     xmm0,dword ptr [__real@425622d1 (07FF6901E16E8h)]  
00007FF68FF04E11  jp          TestSerialization::testStreamRead+0AC5h (07FF68FF05475h)  
00007FF68FF04E17  jne         TestSerialization::testStreamRead+0AC5h (07FF68FF05475h)  
```

</details>

Also checked the Debug build, it did not inline functions but did the same division by `1000.`.

**Comment by asrelo** [comment]
Dunno how this can be fixed, maybe special cases for compilation of particular files...

If ignoring it as insignificant, it would then make sense to remove (or disable in certain builds) parts of the failing unit test.

**Comment by asrelo** [comment]
(MSVC has pragmas to control FP optimizations locally, but I can't see a single use of `#pragma` in code save for `#pragma once`.)

**Comment by appgurueu** [comment]
Spurious macOS CI failure, probably a similar issue: https://github.com/luanti-org/luanti/actions/runs/26355250209/job/77580701902

> Test assertion failed: readV3F1000(is) == v3f(500, 10024.2f, -192.54f)
    at test_serialization.cpp:275

**Comment by SmallJoker** [comment]
This is an old floating point rounding issue in a new skin. https://github.com/luanti-org/luanti/issues/2231#issuecomment-204299324

In the end we might just have to `std::fabs(read_value - expected_value) < 0.0001f`.

**Comment by asrelo** [comment]
I struggle to comprehend a serialize-deserialize roundtrip within one machine failing to reproduce _a number_.

**Comment by appgurueu** [comment]
> I struggle to comprehend a serialize-deserialize roundtrip within one machine failing to reproduce a number.

Well, that's because it isn't a roundtrip test :P

The istream is constructed from `test_serialized_data`, which is a binary blob hardcoded in hex form. This makes some sense: Because this is used for networking, you want it to produce the same result on different machines, and you want it to correctly read what an older peer produced.

F1000 is used only for object position, velocity and rotation (sent over the network), as well as node timer elapsed time and timeout (not sent over the network). I imagine small errors for those are essentially irrelevant, especially as somewhat imprecise math that will vary from platform to platform is then done on each of those.

TL;DR: Approximate comparison is the way to go. I would suggest a relative tolerance on the order of a few ULP rather than an absolute tolerance.
